File: NC-150/07_Trees/47_maximum_depth_binary_tree.py
"""
Problem: Maximum Depth of Binary Tree (LeetCode 104)

Given the root of a binary tree, return its maximum depth.
Maximum depth is the number of nodes along the longest path from root to leaf.

Time Complexity: O(n) - visit every node once
Space Complexity: O(h) - recursion stack, where h is height
"""

from typing import Optional
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        """Recursive DFS approach"""
        if not root:
            return 0
        
        left_depth = self.maxDepth(root.left)
        right_depth = self.maxDepth(root.right)
        
        return 1 + max(left_depth, right_depth)
    
    def maxDepthBFS(self, root: Optional[TreeNode]) -> int:
        """Iterative BFS approach - count levels"""
        if not root:
            return 0
        
        queue = deque([root])
        depth = 0
        
        while queue:
            depth += 1
            level_size = len(queue)
            
            for _ in range(level_size):
                node = queue.popleft()
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
        
        return depth
    
    def maxDepthDFS(self, root: Optional[TreeNode]) -> int:
        """Iterative DFS approach using stack"""
        if not root:
            return 0
        
        stack = [(root, 1)]
        max_depth = 0
        
        while stack:
            node, depth = stack.pop()
            max_depth = max(max_depth, depth)
            
            if node.left:
                stack.append((node.left, depth + 1))
            if node.right:
                stack.append((node.right, depth + 1))
        
        return max_depth

# Helper function to build tree
def build_tree(values):
    if not values or values[0] is None:
        return None
    
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    
    while queue and i < len(values):
        node = queue.popleft()
        
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    
    return root

if __name__ == "__main__":
    sol = Solution()
    
    # Test case 1: [3,9,20,null,null,15,7] -> 3
    root1 = build_tree([3, 9, 20, None, None, 15, 7])
    print(f"Tree [3,9,20,null,null,15,7]")
    print(f"Max depth: {sol.maxDepth(root1)} (expected: 3)")
    assert sol.maxDepth(root1) == 3
    
    # Test case 2: [1,null,2] -> 2
    root2 = TreeNode(1)
    root2.right = TreeNode(2)
    print(f"\nTree [1,null,2]")
    print(f"Max depth: {sol.maxDepth(root2)} (expected: 2)")
    assert sol.maxDepth(root2) == 2
    
    # Test case 3: Empty tree -> 0
    root3 = None
    print(f"\nEmpty tree")
    print(f"Max depth: {sol.maxDepth(root3)} (expected: 0)")
    assert sol.maxDepth(root3) == 0
    
    # Test case 4: Single node -> 1
    root4 = TreeNode(1)
    print(f"\nSingle node")
    print(f"Max depth: {sol.maxDepth(root4)} (expected: 1)")
    assert sol.maxDepth(root4) == 1
    
    # Test case 5: Skewed tree -> 4
    root5 = TreeNode(1, TreeNode(2, TreeNode(3, TreeNode(4))))
    print(f"\nSkewed tree (left only)")
    print(f"Max depth: {sol.maxDepth(root5)} (expected: 4)")
    assert sol.maxDepth(root5) == 4
    
    # Test BFS solution
    root6 = build_tree([3, 9, 20, None, None, 15, 7])
    print(f"\nBFS solution: {sol.maxDepthBFS(root6)} (expected: 3)")
    assert sol.maxDepthBFS(root6) == 3
    
    # Test DFS solution
    print(f"DFS solution: {sol.maxDepthDFS(root6)} (expected: 3)")
    assert sol.maxDepthDFS(root6) == 3
    
    print("\nAll test cases passed!")

File: NC-150/07_Trees/47_maximum_depth_binary_tree.cpp
/*
 * Problem: Maximum Depth of Binary Tree (LeetCode 104)
 * 
 * Given the root of a binary tree, return its maximum depth.
 * Maximum depth is the number of nodes along the longest path from root to leaf.
 * 
 * Time Complexity: O(n) - visit every node once
 * Space Complexity: O(h) - recursion stack, where h is height
 */

#include <iostream>
#include <queue>
#include <algorithm>
#include "tree_node.h"

class Solution {
public:
    // Solution 1: Recursive DFS
    // Time: O(n), Space: O(h)
    int maxDepthRecursive(TreeNode* root) {
        if (root == nullptr) return 0;
        
        int leftDepth = maxDepthRecursive(root->left);
        int rightDepth = maxDepthRecursive(root->right);
        
        return 1 + std::max(leftDepth, rightDepth);
    }
    
    // Solution 2: Iterative BFS (level order)
    // Time: O(n), Space: O(w) where w is max width
    int maxDepthBFS(TreeNode* root) {
        if (root == nullptr) return 0;
        
        std::queue<TreeNode*> q;
        q.push(root);
        int depth = 0;
        
        while (!q.empty()) {
            int levelSize = q.size();
            depth++;
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode* node = q.front();
                q.pop();
                
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
        }
        
        return depth;
    }
    
    // Solution 3: Iterative DFS with stack
    // Time: O(n), Space: O(h)
    int maxDepthDFS(TreeNode* root) {
        if (root == nullptr) return 0;
        
        std::stack<std::pair<TreeNode*, int>> stk;
        stk.push({root, 1});
        int maxDepth = 0;
        
        while (!stk.empty()) {
            auto [node, depth] = stk.top();
            stk.pop();
            
            maxDepth = std::max(maxDepth, depth);
            
            if (node->left) stk.push({node->left, depth + 1});
            if (node->right) stk.push({node->right, depth + 1});
        }
        
        return maxDepth;
    }
    
    // Main function using recursive approach
    int maxDepth(TreeNode* root) {
        return maxDepthRecursive(root);
    }
};

int main() {
    Solution sol;
    
    // Test case 1: [3,9,20,null,null,15,7] -> 3
    TreeNode* root1 = buildTree({3, 9, 20, INT_MIN, INT_MIN, 15, 7});
    std::cout << "Tree: [3,9,20,null,null,15,7]" << std::endl;
    std::cout << "Max depth: " << sol.maxDepth(root1) << " (expected: 3)" << std::endl;
    deleteTree(root1);
    
    // Test case 2: [1,null,2] -> 2
    TreeNode* root2 = new TreeNode(1);
    root2->right = new TreeNode(2);
    std::cout << "\nTree: [1,null,2]" << std::endl;
    std::cout << "Max depth: " << sol.maxDepth(root2) << " (expected: 2)" << std::endl;
    deleteTree(root2);
    
    // Test case 3: Empty tree -> 0
    TreeNode* root3 = nullptr;
    std::cout << "\nEmpty tree" << std::endl;
    std::cout << "Max depth: " << sol.maxDepth(root3) << " (expected: 0)" << std::endl;
    
    // Test case 4: Single node -> 1
    TreeNode* root4 = new TreeNode(1);
    std::cout << "\nSingle node tree" << std::endl;
    std::cout << "Max depth: " << sol.maxDepth(root4) << " (expected: 1)" << std::endl;
    deleteTree(root4);
    
    // Test case 5: Skewed tree [1,2,null,3,null,4] -> 4
    TreeNode* root5 = new TreeNode(1);
    root5->left = new TreeNode(2);
    root5->left->left = new TreeNode(3);
    root5->left->left->left = new TreeNode(4);
    std::cout << "\nSkewed tree (left only)" << std::endl;
    std::cout << "Max depth: " << sol.maxDepth(root5) << " (expected: 4)" << std::endl;
    deleteTree(root5);
    
    // Test BFS solution
    TreeNode* root6 = buildTree({3, 9, 20, INT_MIN, INT_MIN, 15, 7});
    std::cout << "\nBFS solution: " << sol.maxDepthBFS(root6) << " (expected: 3)" << std::endl;
    deleteTree(root6);
    
    std::cout << "\nAll test cases passed!" << std::endl;
    return 0;
}

